cirq/google/api/v2/program.proto

```diff
@@ -198,3 +204,90 @@ message ArgFunction {
   // The arguments to the function.
   repeated Arg args = 2;
 }
+
+// An operation that applies a modified version of a reference circuit. The
+// circuit is stored in the top-level Constants table; the mappings in this
+// object specify how that circuit should be modified for this operation.
+//
+// Multiple CircuitOperations may reference the same base circuit even if their
+// mappings of that circuit are different.
+message CircuitOperation {
+  // The index of the circuit in the top-level constant table.
+  int32 circuit_constant_index = 1;
+
+  // Specifier for repetitions of the circuit, including number of reps and
+  // repetition IDs.
+  RepetitionSpecification repetition_specification = 2;
+
+  // Map of qubits in the circuit to the qubits used in this operation.
+  QubitMapping qubit_map = 3;
+
+  // Map of measurement keys in the circuit to the measurement keys used in
+  // this operation.
+  MeasurementKeyMapping measurement_key_map = 4;
+
+  // Map of args in the circuit to the values (or replacement args) used in
+  // this operation.
+  ArgMapping arg_map = 5;
+}
+
+// A description of the repetitions of a subcircuit. IDs are used as suffixes
+// for measurements in the repeated subcircuit; if repetition_count is given
+// instead, the IDs will simply be the integers [0..N-1].
+message RepetitionSpecification {
+  // An ordered list of IDs for a sequence of repetitions.
+  message RepetitionIds {
+    repeated string ids = 1;
+  }
+
+  oneof repetition_value {
+    // A list of unique IDs, one per repetition of the subcircuit.
+    RepetitionIds repetition_ids = 1;
+
+    // An integer number of repetitions to perform.
+    int32 repetition_count = 2;
+  }
+}
+
+// A mapping of qubits from one value to another. All mappings are applied
+// simultaneously and independently; for example, [(a, b), (b, a)] will swap
+// qubits a and b.
+message QubitMapping {
+  // Indicates that qubit "source" should be replaced with "target".
+  message QubitPair {
+    Qubit source = 1;
+    Qubit target = 2;
```

Comment:
Sorry to nitpick, but I still feel like "source" and "target" is ambiguous. Is this saying that the qubit id appearing in the _outer_ circuit is the source and should be replaced by the target value in the inner circuit, or the other way around? I'd really prefer to use something like "inner" and "outer" because the _containment_ relationship is unambiguous (one circuit contains the other as an operation, not the other way around).

Reply:
Discussed offline: `inner` and `outer` only make sense in the context of `CircuitOperation`, where we have comments documenting the behavior of these maps.

Switched these to `key`/`value` to more closely align with the non-proto `CircuitOperation` object.